index.bs

```diff
@@ -7,14 +7,110 @@ Group: browser-testing-tools
+    type: dfn
+        text: WebSocket ws-URI; url: section-3
+        text: A WebSocket Message Has Been Received; url: section-6.2
+        text: The WebSocket Closing Handshake is Started; url: section-7.1.3
+spec: WEBDRIVER; urlPrefix: https://w3c.github.io/webdriver/
+    type: dfn
+        text: additional capability deserialization algorithm; url: dfn-additional-capability-deserialization-algorithm
+        text: additional WebDriver capability; url: dfn-additional-webdriver-capability
+        text: capability name; url: dfn-capability-name
+        text: error code; url: dfn-error-code
+        text: error; url: dfn-errors
+        text: invalid argument; url: dfn-invalid-argument
+        text: matched capability serialization algorithm; url: dfn-matched-capability-serialization-algorithm
+        text: remote end; url: dfn-remote-ends
+        text: session; url: dfn-sessions
+        text: success; url: dfn-success
 </pre>
 
 Introduction {#intro}
 =====================
 
-[[WEBDRIVER|WebDriver]] defines a protocol for introspection and remote control
-of user agents. This specification extends WebDriver by introducing
-bidirectional communication. In place of the strict command/response
-format of WebDriver, this permits events to stream from the user
-agent to the controlling software, better matching the evented nature
-of the browser DOM.
+[[WEBDRIVER|WebDriver]] defines a protocol for introspection and
+remote control of user agents. This specification extends WebDriver by
+introducing bidirectional communication. In place of the strict
+command/response format of WebDriver, this permits events to stream
+from the user agent to the controlling software, better matching the
+evented nature of the browser DOM.
+
+Transport {#transport}
+======================
+
+Message transport is provided using a [[RFC6455|WebSocket]]
+connection.
+
+A [=remote end=] has a <dfn>connection</dfn>, which is a network connection, and
+a <dfn>websocket url</dfn>, which is a string. These are both initially null.
+
+When data is recieved on [=connection=] it must be processed according to the
+requirements of the [[!RFC6455|WebSockets]] specification. When [=a WebSocket
+message has been received=], an implementation must [=handle an incoming
+message=]. When [=the WebSocket closing handshake is started=] an implementation
+must [=end the session=].
+
+When required to <dfn>ensure a connection is established</dfn>, run
+the following steps:
+
+1.  If the [=websocket url=] is not null, return the
+    [=websocket url=]
+
+2.  Set up a network connection that listens on an implementation-defined
+    hostname <var>host</var> and port <var>port</var>. Set the [=connection=] to
+    this network connection.
+
+3.  Let <var>WebSocket URL</var> be the result of constructing a [=WebSocket
+    ws-URI=] with host <var>host</var>, port <var>port</var>, and empty path an
+    query.
+
+4.  Set the [=websocket url=] to <var>WebSocket URL</var>.
```

Comment:
I'm not sure I can predict how the "ensure" aspect of this algorithm will turn out to be useful. If it's not so clear, maybe we can drop it until later when it's needed.

Reply:
Unfortunately it's already needed in a theoretical way. The way that session matching works in WebDriver you can have something like

```
capabilities: {firstMatch:[{bidi: true, browserName: "Some other browser"}, {bidi: true}])
```

In that case you'd go through the capabilities matching steps twice, once for the first entry, which would then fail, and once for the second entry. So because we establish the connection during capabilities matching, there's a spec-theoretic possibility that we previously established a connection for some prior set of capabilities and want to return that. Now, in practice, I don't expect an implementation to actually do that. But it's easier to write the small fiction here than it is to add another hook to allow the capabilities set to be post-processed after selection.

Reply:
OK, so I realised that this doesn't work anyway if we want to tie the URL to a session. So I guess I have to fix it properly.

Reply:
I agree that we shouldn't (can't?) establish a ws connection during [new session](https://w3c.github.io/webdriver/#new-session), at least not if "establish" means making a network connection. Rather, we bind the port and are ready to accept new ws connections, like a ws server.

Having the ws URL be unguessable will probably preserve some flexibility, but we need to return the ws URL somewhere. It's not the capability itself, so I guess either a "websocketURL" capability, or as a sibling to the capabilities in the response. I'm not sure if this detail has already been discussed?

Aside on capabilties: per my testing of current chromedriver, geckodriver and safaridriver, neither `{"capabilities":{"firstMatch":[{"bidi":true},{"bidi":false}]}}` nor `{"capabilities":{"firstMatch":[{"bidi":true},{}]}}` will work, because they all return an error due to `bidi` not being a know capability. If there are clients that could work with and without bidi, they'd have to try connecting twice.

Reply:
We can return the URL in the bidi field of the capability response. There isn't a problem here :)

You're right, the correct new session capabilities would have been `{capabilities: {firstMatch: [ {bidi: true}, {} ] } }`.

Reply:
Note that `{"capabilities":{"firstMatch":[{"bidi":true},{}]}}` actually doesn't work, that's one of the forms I tested. I've filed https://github.com/w3c/webdriver/issues/1528 with a lot more details.

Reply:
If we're going to use the same capability to request bidi and return the ws URL, I'd like to bikeshed the name and suggest "webSocketUrl", because it describes what you're getting and allows for another protocol in some distant future. The existing "proxyAutoconfigUrl" is similarly named, and it's also similar to the [`GET /json/version` endpoint in CDP](https://chromedevtools.github.io/devtools-protocol/#get-jsonversion).

Reply:
@jgraham has renamed it "webSocketUrl"